pkg/debuginfo/store.go

```diff
@@ -192,87 +205,118 @@ func (s *Store) Upload(stream debuginfopb.DebugInfoService_UploadServer) error {
 		}
 	} else {
-		level.Error(s.logger).Log("msg", "failed to fetch metadata state", "err", err)
+		if !errors.Is(err, ErrMetadataNotFound) {
+			level.Error(s.logger).Log("msg", "failed to fetch metadata state", "err", err)
+		}
 	}
 
 	found, err := s.find(ctx, buildID)
 	if err != nil {
 		return err
 	}
 
+	var objFileHash string
 	if found {
+		if req.GetInfo().Hash != "" {
+			if metadataFile != nil && metadataFile.Hash != "" {
+				if metadataFile.Hash == req.GetInfo().Hash {
+					level.Debug(s.logger).Log("msg", "debug info already exists", "buildid", buildID)
+					return status.Error(codes.AlreadyExists, "debuginfo already exists")
+				}
+			}
+		}
+
 		objFile, err := s.fetchObjectFile(ctx, buildID)
 		if err != nil {
 			return status.Error(codes.Internal, err.Error())
 		}
 
-		if req.GetInfo().Hash != "" {
-			h, err := hash.File(objFile)
-			if err != nil {
-				return status.Error(codes.Internal, err.Error())
-			}
-
-			if h == req.GetInfo().Hash {
-				level.Debug(s.logger).Log("msg", "debug info already exists", "buildid", buildID)
-				return status.Error(codes.AlreadyExists, "debuginfo already exists")
+		if err := elfutils.ValidateFile(objFile); err != nil {
+			// Failed to validate. Mark the file as corrupted, and let the client try to upload it again.
+			if err := s.metadataManager.corrupted(ctx, buildID); err != nil {
+				level.Warn(s.logger).Log("msg", "failed to update metadata as corrupted", "err", err)
 			}
+			// Client will retry.
+			return status.Error(codes.Internal, err.Error())
 		}
 
+		// Valid.
 		hasDWARF, err := elfutils.HasDWARF(objFile)
-		if err != nil && !errors.Is(err, io.EOF) {
-			var fe *elf.FormatError
-			if errors.As(err, &fe) {
-				// Ignore bad magic number if all zero
-				if !strings.Contains(fe.Error(), "bad magic number '[0 0 0 0]'") {
-					return status.Error(codes.Internal, err.Error())
-				}
-			} else {
-				return status.Error(codes.Internal, err.Error())
-			}
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "failed to check for DWARF", "err", err)
 		}
 		if hasDWARF {
-			// We probably have the best version.
-			level.Debug(s.logger).Log("msg", "debug info with DWARF already exists", "buildid", buildID)
 			return status.Error(codes.AlreadyExists, "debuginfo already exists")
 		}
 	}
 
-	if err := s.metadataManager.update(ctx, buildID, metadataStateUploading); err != nil {
-		level.Error(s.logger).Log("msg", "failed to update metadata", "err", err)
+	if err := s.metadataManager.uploading(ctx, buildID); err != nil {
+		err := fmt.Errorf("failed to update metadata before uploading: %w", err)
+		return status.Error(codes.Internal, err.Error())
 	}
 
-	// At this point we know that we still have a better version of the debug information file,
+	//nolint:forcetypeassert
+	b := s.pool.Get().([]byte)
+	buf := flexbuf.With(b)
+	defer func() {
+		// TODO(kakkoyun): Check if this creates any problems with reused buffers.
```

Comment:
Which kind of problems are you anticipating?

Reply:
I haven't properly checked if we reuse the byte slices underneath for these buffers. Because the wrapper buffers don't check boundaries while reading, they read until the end of the []byte underneath. This could be a problem if we got a reused []byte that is larger than the current data. I might be overthinking and missing things in here. That's why I want to double-check before merging.

Reply:
Make sense, thanks for explaining!! I think that for this part, it would be good to have someone with actual Go expertise to take a look 😄